webapp/application.py

```diff
@@ -119,10 +132,172 @@ def application_page(token):
+        "careers/application/index.html",
         title="You application",
+        withdrawal_reasons=withdrawal_reasons,
         candidate=candidate,
         application=application,
         job=job,
         hiring_lead=hiring_lead,
+        token=token,
+    )
+
+
+@application.route("/withdraw/<string:token>")
+def application_withdrawal(token):
+    payload = cipher.decrypt(token)
+
+    payload = json.loads(payload)
+    withdrawal_reason_id = payload.get("withdrawal_reason")
+    candidate_id = payload["candidate_id"]
+    application_id = payload["application_id"]
+
+    application = harvest.get_application(application_id)
+    job = harvest.get_job(application["jobs"][0]["id"])
+
+    hiring_lead = None
+    for recruiter in job["hiring_team"]["recruiters"]:
+        if recruiter["responsible"]:
+            hiring_lead = harvest.get_user(recruiter["id"])
+            break
+
+    if (
+        "email" not in payload
+        or application.get("candidate_id") != candidate_id
+    ):
+        flask.abort(403)
+
+    withdrawal_reason = withdrawal_reasons.get(withdrawal_reason_id)
+    hiring_lead_id = hiring_lead["id"]
+
+    # call the Harvest API to reject the application
+    response = harvest.reject_application(
+        application_id, hiring_lead_id, withdrawal_reason_id, withdrawal_reason
+    )
+    response.raise_for_status()
+
+    return flask.render_template("careers/application/withdrawal.html")
+
+
+@application.route("/withdraw/<string:token>", methods=["POST"])
+def sendForm(token):
+    decrypted = cipher.decrypt(token)
+    if not decrypted:
+        flask.abort(404)
+    name, candidate_id, application_id = tuple(decrypted.split("-"))
+    candidate_id, application_id = int(candidate_id), int(application_id)
+    application = harvest.get_application(application_id)
+    candidate = harvest.get_candidate(application["candidate_id"])
+    job = harvest.get_job(application["jobs"][0]["id"])
+
+    for recruiter in job["hiring_team"]["recruiters"]:
+        if recruiter["responsible"]:
+            hiring_lead = harvest.get_user(recruiter["id"])
+            break
+
+    if (
+        "candidate_id" not in application
+        or application["candidate_id"] != candidate_id
+        or not candidate["first_name"].lower() == name
+    ):
+        flask.abort(404)
+
+    candidate_name = candidate["first_name"]
+    position = application["jobs"][0]["name"]
+    hiring_lead = hiring_lead["name"]
+
+    # Sanitize and parse user input
+    email = parseaddr(flask.request.form["email"])[1]
+    candidate_email = parseaddr(candidate["email_addresses"][0]["value"])[1]
+
+    if flask.request.form["withdrawal-reason-other"]:
+        withdrawal_reason = flask.request.form["withdrawal-reason-other"]
+    else:
+        withdrawal_reason = flask.request.form["withdrawal-reason"]
+
+    # Reject if user typed the wrong email
+    if candidate_email != email:
+        return redirect(flask.request.referrer + "#wrong-email")
```

Comment:
@nottrobin Would it be better to show a new page as well when a wrong email is typed? 
This works but one issue is the notification message is hidden by navigation even though the hash url points the matching id section. Would there be a better way for this?

Reply:
Yes this is why I wasn't crazy about the notification solution. It feels like it can easily go wrong and the message doesn't get through.

But I'm gonna fix the other thing first and then think about this. I'll add a comment at the bottom about my conversation with Marina.